src/Microsoft.AspNet.Razor.Runtime/Runtime/TagHelpers/TagHelperDesignTimeDescriptorFactory.cs

```diff
@@ -90,31 +93,55 @@ public static TagHelperDesignTimeDescriptor CreateDescriptor(Type type)
             return null;
         }
 
-        private static DocumentationDescriptor CreateDocumentationDescriptor(Assembly assembly, string id)
+        private XmlDocumentationProvider GetXmlDocumentationProvider(Assembly assembly)
         {
-            var assemblyLocation = assembly.Location;
+            var hashCodeCombiner = HashCodeCombiner.Start();
+            hashCodeCombiner.Add(assembly);
+            hashCodeCombiner.Add(CultureInfo.CurrentCulture);
+            var cacheKey = hashCodeCombiner.CombinedHash;
 
-            if (string.IsNullOrEmpty(assemblyLocation) && !string.IsNullOrEmpty(assembly.CodeBase))
+            XmlDocumentationProvider documentationProvider = null;
+            lock (_cacheLock)
             {
-                var uri = new UriBuilder(assembly.CodeBase);
+                if (!_documentationProviderCache.TryGetValue(cacheKey, out documentationProvider))
+                {
+                    var assemblyLocation = assembly.Location;
 
-                // Normalize the path to a UNC path. This will remove things like file:// from start of the uri.Path.
-                assemblyLocation = Uri.UnescapeDataString(uri.Path);
-            }
+                    if (string.IsNullOrEmpty(assemblyLocation) && !string.IsNullOrEmpty(assembly.CodeBase))
+                    {
+                        var uri = new UriBuilder(assembly.CodeBase);
 
-            // Couldn't resolve a valid assemblyLocation.
-            if (string.IsNullOrEmpty(assemblyLocation))
-            {
-                return null;
+                        // Normalize the path to UNC path. This will remove things like file:// from start of uri.Path.
+                        assemblyLocation = Uri.UnescapeDataString(uri.Path);
+                    }
+
+                    // Couldn't resolve a valid assemblyLocation.
+                    if (string.IsNullOrEmpty(assemblyLocation))
+                    {
+                        return null;
+                    }
+
+                    var xmlDocumentationFile = GetXmlDocumentationFile(assembly, assemblyLocation);
```

Comment:
Is this operation expensive? If so, maybe a `ConcurrentDictionary` would be a better alternative to the lock (since you wouldn't block all reads while you're constructing the documentation)